### ISSUE TITLE
Allow users to use multiple auth providers

### ISSUE THREAD

**Issue by abdonrd** [strapi/strapi#2468]
- ❌ **I have created my request on the Product Board before I submitted this issue**
- ✅ **I have looked at all the other requests on the Product Board before I submitted this issue**

**Please describe your feature request:**

Allow users to use multiple auth providers.

Example: I want to have the Google & Twitter auth providers linked with my user.

**Comment by derrickmehaffy** [comment]
I think this will get easier once the authentication plugins are rewritten in passport but will have to think about how we can do this. :thinking:

**Comment by lauriejim** [comment]
Thank you @abdonrd for this feature request.

**Comment by z-ax** [comment]
So, any estimates on this?
Or, maybe, some spec/direction on implementation?

**Comment by ishan123456789** [comment]
This need more attention for sure it's hard to force the user to use the same provider they used earlier. Is there any hack available.

**Comment by ishan123456789** [comment]
@lauriejim Can you please review this Gist I've updated the provider service a bit to allow user to use multiple providers for different providers

https://gist.github.com/ishan123456789/2fba00d48c0d7d900d03d97d35c90c4e

**Comment by rochadsouzaiit** [comment]
I think this is a key feature. Could you (anyone from Strapi team) validate the gist above, please ☝️?  Ty 🙏
Btw, why is this issue closed?

**Comment by z-ax** [comment]
Also don't know why it is closed, didn't see it implemented.

**Comment by oskar-gmerek** [comment]
I was sure that feature is implemented from the scratch. Shock, its not. Its not even planed yet :(

**Comment by dmythro** [comment]
Was it an accidental ticket closing?

**Comment by derrickmehaffy** [comment]
@dmythro to keep issues clean, once they are added to the product board the feature request is closed here but kept unlocked to allow for discussion. We don't use Github issues to track the progress of a feature request, that is all handled on our product board.

**Comment by lauriejim** [comment]
Hello @ishan123456789 , it looks good.
You removed the `provider` constraint so that's good.

**Comment by borm** [comment]
@abdonrd @lauriejim any updates on it?

**Comment by derrickmehaffy** [comment]
No updates, when we start work on this you will see the card move in productboard.

**Comment by var-bin** [comment]
Hi there,

I also bumped into the same issue. I found a Gist by @ishan123456789. It looks like there are some additional steps in order to get it to work. Do you happen to know what additional steps need to be done in this case?

Cheers,
Vitalii

**Comment by derrickmehaffy** [comment]
> Hi there,
> 
> I also bumped into the same issue. I found a Gist by @ishan123456789. It looks like there are some additional steps in order to get it to work. Do you happen to know what additional steps need to be done in this case?
> 
> Cheers,
> Vitalii

We store the provider in a string field and not an array along with the requirements of having a unique key (email I believe) means we can't have multiple "users", so we need a way to store the provider information on a single entry. This is something we will probably look into when we get around to refactoring the users-permissions plugin.

That refactoring will most likely happen when we develop the native API token feature as it will become the default auth system in Strapi with the users-permissions plugin becoming optional.

https://portal.productboard.com/strapi/1-roadmap/c/40-api-token-with-permissions

**Comment by var-bin** [comment]
Hi @derrickmehaffy,

Thanks a lot for your answer. It's not about multiple users but about multiple providers for one user account. To make a long story short, I was able to implement this feature for our needs after exploring Strapi dev docs and Gist from @ishan123456789. Cheers

**Comment by derrickmehaffy** [comment]
> Hi @derrickmehaffy,
> 
> Thanks a lot for your answer. It's not about multiple users but about multiple providers for one user account. To make a long story short, I was able to implement this feature for our needs after exploring Strapi dev docs and Gist from @ishan123456789. Cheers

Yeah I'm aware, but we need to store the provider information on the user (for multiple providers) and currently an array field type doesn't exist that can do that. We are looking into an array type field in our v4 development combined with our need to refactor the U&P plugin.

**Comment by var-bin** [comment]
> > Hi @derrickmehaffy,
> > Thanks a lot for your answer. It's not about multiple users but about multiple providers for one user account. To make a long story short, I was able to implement this feature for our needs after exploring Strapi dev docs and Gist from @ishan123456789. Cheers
> 
> Yeah I'm aware, but we need to store the provider information on the user (for multiple providers) and currently an array field type doesn't exist that can do that. We are looking into an array type field in our v4 development combined with our need to refactor the U&P plugin.

OK, got you. Thanks a bunch for your attention to this matter

**Comment by otaviocx** [comment]
Hi guys! Since it is already closed here and there is no information about planning and the estimated delivery date in the Product Board for that, I was wondering where are you tracking this request and putting such planning? Thanks in advance!

**Comment by derrickmehaffy** [comment]
> Hi guys! Since it is already closed here and there is no information about planning and the estimated delivery date in the Product Board for that, I was wondering where are you tracking this request and putting such planning? Thanks in advance!

No not currently, there was some attempts in the past to get it working (one attempt was by me) but it's not in our planning at this time. If someone else wanted to pick this up and make a PR they are welcome to.

**Comment by TheMangoTrain** [comment]
On this topic, how about simply being able to _change_ providers?

**Comment by bayramorhan** [comment]
any updates?

**Comment by oojroo** [comment]
I would like to provide a PR for that using any provider (external or local) to sign in.
If the email used at the providers are equal and unique_email is active the provider attribute will be ignored while searching the reflective user for sign in.

Can we reopen this issue that i can create a PR?

**Comment by marcelobaez** [comment]
Anyone getting an 403 when making the call to {{strapiURL}}/api/auth/{provider}/callback?access_token=${account.access_token} ? for me theres no way to get strapi creating the user from the callback. And yes, the access token is there

**Comment by galaxxyz** [comment]
@marcelobaez check users permissions in Strapi admin (settings > roles) most likely some routes are not enabled, 403 indicates that the server refuses to authorise your requests

**Comment by bayramorhan** [comment]
Please upvote https://feedback.strapi.io/security/p/multiple-authentication-providers-per-user . this is still a major problem.
